### PR TITLE
fix(floating-menu): positioning in scrollable container maintained

### PR DESCRIPTION
**Pull request by psachs21**
Override style of default render to position offscreen to the right
to prevent the window.scrollY from being affected right before
measuring the size of the menu.

This seems to only be an issue in the latest version of react (16.3) probably
due to the asynchronous rendering changes.

### CODE REVIEW

**Comment by asudoh** [comment]
Thanks alot @psachs21 for finding the issue and trying to fix! The fix signals that we need to better understand the problem ourselves. Meanwhile, if you have more details on your findings it’ll very nice. Thanks!

**Comment by psachs21** [comment]
@asudoh the issue is pretty simple: when the menu first renders, it's rendered as is (without position: absolute) in the regular flow of the document. So depending on the layout of the document, it may adjust the scrollY position (if there are enough elements). This was theoretically a bug before react 16.3, but I suspect was only exposed due to async rendering changes in core React. I think the fix of positioning the element of screen is the easiest solution.

File: src/internal/FloatingMenu.js [hunk 1]
```diff
@@ -318,21 +318,26 @@ class FloatingMenu extends React.Component {
...
-              top: `${pos.top}px`,
-              position: 'absolute',
-              right: 'auto',
-              margin: 0,
-              opacity: 1,
-            },
-            styles
-          ),
-        });
+    const style = pos
```
Comment (joshblack) [inline src/internal/FloatingMenu.js]: As a code style comment, could we re-write this in a way that could communicate why these values are needed? Comments are fine as well 🙂

As an added note, why in this ternary do we apply `styles` when `pos` is defined but not when it isn't defined? Trying to catch up on this component implementation 😄

Comment (psachs21) [reply src/internal/FloatingMenu.js]: @joshblack updated. Should be a little clearer now.

I was thinking I didn't want to apply stylings if we're just hiding it off to the side but you're right, if those stylings change the size of the element, that should be included. Updated.

**Comment by psachs21** [comment]
@asudoh I've also been able to reproduce the issue in the playground, you just need to set some css:

```css
html, body, #root, [role='main'] {
  height: 100%;
}
```

And make sure your example story has a top level div that also has height: 100%.

I haven't submitted an update to the story because I don't think it serves much of a purpose other than to demonstrate the bug.

File: src/internal/FloatingMenu.js [hunk 2]
```diff
@@ -318,21 +318,27 @@ class FloatingMenu extends React.Component {
...
+    // If no pos available, we need to hide the element (offscreen to the left)
+    // This is done so we can measure the content before positioning it correctly.
+    const positioningStyle = pos
+      ? {
+          left: `${pos.left}px`,
+          top: `${pos.top}px`,
+          right: 'auto',
+        }
+      : {
+          right: `${window.innerWidth}px`,
```
Comment (asudoh) [inline src/internal/FloatingMenu.js]: Thanks @psachs21 for explaining your finding in detail! Your fix makes sense basically, just one thing - Given you want to make the temporally menu location absolute and off-screen - Did you mean `left` here?

Comment (psachs21) [reply src/internal/FloatingMenu.js]: It's currently positioned off the the left of the screen, but we can put it off to the right as well. Both are roughly equivalent.

Comment (asudoh) [reply src/internal/FloatingMenu.js]: Yeah I thought the same - Like `right: `${window.innerWidth}px` or `left: 0`.

Comment (asudoh) [reply src/internal/FloatingMenu.js]: ^ I guess we'd choose either of those. Thanks!

Comment (psachs21) [reply src/internal/FloatingMenu.js]: hmm, your right, I had a brain fart. Left would be the natural fit for me. I'll update the PR.

**Pull request merged**